### ISSUE TITLE
3.5" FDD controller support (selectable in slot 5, 6 or 7)

### ISSUE THREAD

**Issue by tomcw** [AppleWin/AppleWin#104]
**BerliOS Feature Request #4879**

**Date:** 2009-November-13 22:39
**Submitted By:** tomch
**Category:** General Functionality
**Assigned To:** tomch
**Priority:** 5
# 

Date: 2009-Nov-13 22:39
Sender: tomch
Logged In: YES 
user_id=25188
Browser: Mozilla/5.0 (Windows; U; Windows NT 5.1; en-GB; rv:1.9.0.7) Gecko/2009021910
Firefox/3.0.7 (.NET CLR 3.5.30729)

Requested by Stanislav Georgiev (3/11/2009):

3.5" FDD controller support (selectable in slot 5, 6 or
7) for ProDOS

**Comment by tomcw** [comment]
From cea2, "AppleWin 1.26.1.1 Screen Flicker" (22 Apr 2017):

James Davis:
>Also, can you put a pair of 3.5" flippy disk drives in slot 5 specifically for 800KB disk images?--[Like the 5.25" floppy disk drives in slot 6 or the hard disk drives in slot 7.] 

Marco Verpelli:
> +1 for the 3.5" in slot 5

**Comment by SEGStriker** [comment]
Still no news on this topic?

**Comment by tomcw** [comment]
Nothing more.

My time on this AppleWin project is very limited, and the little time I do have has been spent on other issues.

I hope to put out a new release very soon (but not including this item), then I want to review outstanding issues with a view to identifying those with a high vote count, like this one.

**Comment by 8bit-Dude** [comment]
I definetely vote on this one!!
I am working on a cross-platform game for the uthernet card (see: http://www.applefritter.com/?q=content/8bit-slicks-online-top-down-racer-appleataric64)
But 140 KB is not enough space to store all the maps, so 800 KB supported would be awesome!

**Comment by tomcw** [comment]
@8bit-Dude - you do realise that you can use the hybrid (HDD/3.5") controller in slot-7 which supports both HDD (up to 32MB) and 3.5" (800K) images?

**Comment by 8bit-Dude** [comment]
Is it possible to boot on an 800K image?
I tried converting some 800K DSK files to HDV format with CiderPress, but nothing happens... Just the Apple II logo is shown in AppleWin.

**Comment by sicklittlemonkey** [comment]
Go into the config window, Disk tab, enable the HDD controller in slot 7,
and click on the HDD1 drop-down and choose Select Hard Disk Image...

Choose your file. It just needs to be .PO (ProDOS Order).

Not as convenient as 5.25 yet, but it works.

Cheers,
Nick.

**Comment by 8bit-Dude** [comment]
Ah yes, it worked! Thanks for the tip, I will probably distribute my game as a single PO, or two 140K DSK.
But still, would be nice to support 800K DSK in future!

**Comment by 8bit-Dude** [comment]
Just wondered what is the plan on this? The issue has not been closed, so it must mean that you are considering doing something to allow loading 800K disks from the floppy slots?

**Comment by tomcw** [comment]
Hi, I noticed this from your old 2018 comment didn't get a response:
>But still, would be nice to support 800K DSK in future!

As Nick said previously, this is already supported, via the hybrid (HDD/3.5") controller in slot-7.
EG. This [Prince of Persia](https://mirrors.apple2.org.za/ftp.apple.asimov.net/images/games/action/prince_of_persia/Prince%20of%20Persia%20%28Cracked%203.5%20floppy%20for%20IIc%2B%29.po) image will boot if you enable a Harddisk Controller (in a slot-7) and then insert that image.

So you question "to allow loading 800K disks from the floppy slots" - well this is already supported. Just not in slots 5 and 6 (only in slot 7).

**Comment by 8bit-Dude** [comment]
> So you question "to allow loading 800K disks from the floppy slots" - well this is already supported. Just not in slots 5 and 6 (only in slot 7).

Yes I know, but my question is whether it will be supported in slots 5/6 in future, as they are more convenient.
They have icons on main screen, and it is possible to load a slot 5/6 floppy from command line.

**Comment by SEGStriker** [comment]
Historically 800K drives were served by special adapter in slot 5. I think there were advanced (combined) FDD & HDD adapters for slot 7, just because it is with higher priority, otherwise these controllers should work in most other slots as well. As a standard slot 6 was preserved for 140K FDDs, I haven't seen others, but there should be some non-standard FDD controllers that have more than 2 FDD connectors, some of them even for 800K but they are mapped to slot 5 again.
Not sure how PC Transporters FDDs are visible while in A2 mode - under ProDOS, CP/M, as I never had such. But there are some docs that state these 360K FDDs are visible under ProDOS.
So there are lot of options and we are already short of slots... My own Pravetz 8A machine was full loaded, all slots, 2112 KB RAM  on AUX slot and I had 4-5 more adapters that I used on demand.

**Comment by fadden** [comment]
> As Nick said previously, this is already supported, via the hybrid (HDD/3.5") controller in slot-7.

"MUST BOOT FROM SLOT 4, 5 OR 6" -- Apple Pascal 1.3, when booted from an 800KB disk

OTOH, CP/AM 5.1 (the AE version of CP/M) boots happily off of an 800KB image.  The 3.5" disk shows up as both A: and B:, but maybe that's normal?

(AppleWin v1.30.15)

**Comment by tomcw** [comment]
@fadden - there's support (cmd line only) for an HDD or 800K disk controller in slot 5. See the manual or the rel note here:
https://github.com/AppleWin/AppleWin/releases/tag/v1.30.14.0

**Comment by fadden** [comment]
`AppleWin.exe -l -s5 hdc -s5h1 Pascal800.po` stalls at startup.
`AppleWin.exe -l -s7 hdc -s7h1 Pascal800.po` gives me the "must boot from slot 4, 5 or 6" message.

This is with all disks ejected and the slot 7 hard drive disabled in the "disks" GUI.

If the disk image is also configured in the GUI as a hard drive, the emulator fails to load the disk image and tells me to see the log for the reason why.  The log doesn't have an obviously useful message though; just `Init: DoHardDiskInsert(HDD1), res=1` vs. `res=0`.  (I suspect it's trying to open the same file twice with sharing disabled, and getting rejected.)  This came up because, when launched with the -s7/-s7h1 options, the disk image is remembered in the settings.

I finally realized that what I needed to do was:
 - Issue the -s5 -s5h1 command as above.
 - Hit Ctrl+F2 to halt the boot process, which was getting hung up on slot 6.
 - type "PR#5".

Apparently I've spent too much time on a IIgs. :-)  Once I figured that out, it worked fine.

**Comment by tomcw** [comment]
Glad you figured it out.
(I will update the docs with this advice)

Btw, you could optimise the config by *additionally* including `-s6 empty` to get it to auto-boot from slot 5.

**Comment by fadden** [comment]
A few more notes...

`-s6 empty` is "sticky".  You must issue `-s6 diskii` to re-enable the drive, as there is currently no GUI option for this.  Once this was done, Pascal would start to boot and then hang (on a 140KB disk in slot 6).  ProDOS worked fine.

I looked through the documentation but couldn't figure out how to clear settings.  Issue #918 explained it. 
 With the settings reset, booting Pascal off of a 140KB disk worked.

I determined that Pascal 1.1 won't finish booting from a 140KB disk if `-s5 hdc` has been set.  You need to clear it with `-s5 empty` to get it to work again.  Pascal 1.3 does work, so this appears to be an issue with older versions of Pascal encountering something in slot 5 that causes it to hang.

My suggestion would be to either implement the GUI for the options, or don't store those values in the registry.  My expectation when providing command-line arguments for otherwise inaccessible options is that they're only in effect for the duration of that command; having them be "permanent" was confusing because there's no way to see the configuration state in the GUI.

BTW, having `-s5 hdc` set prevents older versions of AppleWin (e.g. 1.30.12) from starting.

**Comment by tomcw** [comment]
Thanks. Much useful feedback and suggestions there. I'll take some action on them (ie. as a minimum update the docs). 

Improving the Config GUI (#853) is long overdue and will address many other outstanding points.
